# Face recognition of 5celebs using TensorFlow Backend

In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os
import keras
import matplotlib.pyplot as plt
from keras.layers import Dense,GlobalAveragePooling2D
from keras.applications import MobileNet
from keras.preprocessing import image
from keras.applications.mobilenet import preprocess_input
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Model
from keras.optimizers import Adam

Using TensorFlow backend.


In [2]:
base_model=MobileNet(weights='imagenet',include_top=False) #imports the mobilenet model and discards the last 1000 neuron layer.

x=base_model.output
x=GlobalAveragePooling2D()(x)
x=Dense(1024,activation='relu')(x) #we add dense layers so that the model can learn more complex functions and classify for better results.
x=Dense(1024,activation='relu')(x) #dense layer 2
x=Dense(512,activation='relu')(x) #dense layer 3
preds=Dense(3,activation='softmax')(x) #final layer with softmax activation

C:\Users\md sohaib uddin\AppData\Local\Continuum\anaconda3\lib\site-packages\keras_applications\mobilenet.py:207: UserWarning: `input_shape` is undefined or non-square, or `rows` is not in [128, 160, 192, 224]. Weights for input shape (224, 224) will be loaded as the default.
  warnings.warn('`input_shape` is undefined or non-square, '


17227776/17225924 [==============================] - 79s 5us/step


In [3]:
model=Model(inputs=base_model.input,outputs=preds)
#specify the inputs
#specify the outputs
#now a model has been created based on our architecture

In [4]:
for layer in model.layers[:20]:
    layer.trainable=False
for layer in model.layers[20:]:
    layer.trainable=True

In [5]:
train_datagen=ImageDataGenerator(preprocessing_function=preprocess_input) #included in our dependencies

train_generator=train_datagen.flow_from_directory('./train/', # this is where you specify the path to the main data folder
                                                 target_size=(224,224),
                                                 color_mode='rgb',
                                                 batch_size=32,
                                                 class_mode='categorical',
                                                 shuffle=True)

Found 61 images belonging to 5 classes.


In [33]:
model.compile(optimizer='Adam',loss='categorical_crossentropy',metrics=['accuracy'])
# Adam optimizer
# loss function will be categorical cross entropy
# evaluation metric will be accuracy

step_size_train=train_generator.n//train_generator.batch_size
model.fit_generator(generator=train_generator,
                   steps_per_epoch=step_size_train,
                   epochs=5)

Epoch 1/5
24/24 [==============================] - 19s 802ms/step - loss: 0.6303 - acc: 0.9114
Epoch 2/5
24/24 [==============================] - 14s 591ms/step - loss: 0.5115 - acc: 0.7967
Epoch 3/5
24/24 [==============================] - 14s 594ms/step - loss: 0.2509 - acc: 0.9240
Epoch 4/5
24/24 [==============================] - 14s 599ms/step - loss: 0.2698 - acc: 0.9375
Epoch 5/5
24/24 [==============================] - 14s 603ms/step - loss: 0.2309 - acc: 0.9427


## Loss Function = 0.2309
## Accuracy = 94.27%

Loss Function used here is Categorical crossentropy
Categorical crossentropy is a loss function that is used for single label categorization. 
This is when only one category is applicable for each data point. In other words, an example can belong to one class only.

